### ISSUE TITLE
Hyphen character in project name creates invalid namespaces

### ISSUE THREAD

**Issue by sandcastle** [OmniSharp/generator-aspnet#256]
When creating a project name using `yo aspnet` if you specify hyphens, its creates namespaces that contain hyphens which will result in the project being unable to build. Ideally the project name should be sanitised when mapping to the namespace to use for the project.

**Comment by peterblazejewicz** [comment]
Maybe a solution like this one is required:
https://github.com/yeoman/generator-angular/blob/master/app/index.js#L15
and tests.

**Comment by peterblazejewicz** [comment]
@sandcastle Thanks, we will check this!

**Comment by sandcastle** [comment]
Happy to send a PR if there is acceptance for taking a dependency on `underscore.string` which is used by angular generator above.

**Comment by peterblazejewicz** [comment]
@sandcastle cool, 
I've found a trait with official VS project name requirements ($safeprojectname$):
https://msdn.microsoft.com/en-us/library/eehb4faa.aspx
I think Sayed can know something about this :)
/cc @sayedihashimi

**Comment by sayedihashimi** [comment]
When you use Visual Studio to create projects $safeprojectname$ is used to ensure that a valid project name is created. It will replace characters like `-` to `_`. I looked up the code that is used to handle this and pasted it below.

```csharp
static private string MakeNameCompliant(string name)
{
	if (String.IsNullOrEmpty(name))
	{
		return "";
	}
	if (Char.IsDigit(name[0]))
		name = "_" + name;
	for (int i = 0; i < name.Length; i++)
	{
		UnicodeCategory cat = char.GetUnicodeCategory(name[i]);

		//TODO: For VC, we need the same method for making the name compliant, except we need to remove the check for '.' below.
		if ((cat != UnicodeCategory.UppercaseLetter) && (cat != UnicodeCategory.LowercaseLetter) &&
			(cat != UnicodeCategory.OtherLetter) && (cat != UnicodeCategory.ConnectorPunctuation) &&
			(cat != UnicodeCategory.ModifierLetter) && (cat != UnicodeCategory.NonSpacingMark) &&
			(cat != UnicodeCategory.SpacingCombiningMark) && (cat != UnicodeCategory.TitlecaseLetter) &&
			(cat != UnicodeCategory.Format) && (cat != UnicodeCategory.LetterNumber) &&
			(cat != UnicodeCategory.DecimalDigitNumber) &&
			(name[i] != '.') && (name[i] != '_'))
		{
			name = name.Replace(name[i], '_');
		}
	}
	return name;
}
```

It we can do the same in yo that would be nice.

**Comment by peterblazejewicz** [comment]
@sayedihashimi Thanks!

**Comment by Spinarooni** [comment]
Unicode categories are not directly supported in Node.  Would you need some sort of package like [XRegExp](https://github.com/slevithan/XRegExp) to help replicate this code?

**Comment by peterblazejewicz** [comment]
@Spinarooni I think we could make most-common cases from a code Sayed posted, so using _/underscore.strings methods will cover most cases. We could even use existing module first, like this one (which uses RegExes):
https://github.com/parshap/node-sanitize-filename
and then apply underscore.strings methods.
Thoughts?

**Comment by sayedihashimi** [comment]
We don't have to use a "direct port" of the code above, but anything that works for most cases should be fine. The most important bit is to make sure projects are created with names that build by default. I think we should be able to cover this with a simple regex. You guys know node much better than I do.

**Comment by peterblazejewicz** [comment]
closed via: e431caa9
